private/bufpkg/bufconfig/write.go

```diff
@@ -273,30 +257,162 @@ func writeConfig(
 	for _, option := range options {
 		option(writeConfigOptions)
 	}
-	if !writeConfigOptions.documentationComments && writeConfigOptions.uncomment {
-		return errors.New("cannot set uncomment without documentationComments for WriteConfig")
+	if err := validateWriteConfigOptions(writeConfigOptions); err != nil {
+		return err
 	}
-	if writeConfigOptions.moduleIdentity == nil && len(writeConfigOptions.dependencyModuleReferences) > 0 {
-		return errors.New("cannot set deps without a name for WriteConfig")
+	// This is the same default as the bufconfig getters.
+	version := V1Beta1Version
+	if writeConfigOptions.version != "" {
+		if err := validateVersion(writeConfigOptions.version); err != nil {
+			return err
+		}
+		version = writeConfigOptions.version
 	}
-	externalConfigV1 := ExternalConfigV1{
-		Version: V1Version,
+	config := &Config{
+		Version:        version,
+		ModuleIdentity: writeConfigOptions.moduleIdentity,
 	}
-	externalConfigV1.Lint.Use = defaultLintIDs
-	externalConfigV1.Breaking.Use = defaultBreakingIDs
-	if writeConfigOptions.moduleIdentity != nil {
-		externalConfigV1.Name = writeConfigOptions.moduleIdentity.IdentityString()
+	if writeConfigOptions.breakingConfig != nil {
+		config.Breaking = writeConfigOptions.breakingConfig
+		// Always force the provided version.
+		config.Breaking.Version = version
 	}
-	for _, dependencyModuleReference := range writeConfigOptions.dependencyModuleReferences {
-		externalConfigV1.Deps = append(
-			externalConfigV1.Deps,
-			dependencyModuleReference.String(),
-		)
+	if writeConfigOptions.lintConfig != nil {
+		config.Lint = writeConfigOptions.lintConfig
+		// Always force the provided version.
+		config.Lint.Version = version
+	}
+	var dependencies []string
+	if len(writeConfigOptions.dependencyModuleReferences) > 0 {
+		dependencies = make([]string, len(writeConfigOptions.dependencyModuleReferences))
+		for i, dependencyModuleReference := range writeConfigOptions.dependencyModuleReferences {
+			dependencies[i] = dependencyModuleReference.String()
+		}
 	}
-	tmplData := tmplUndocumentedData
 	if writeConfigOptions.documentationComments {
-		tmplData = tmplDocumentationCommentsData
+		// Write out config using template with document comments.
+		return writeCommentedConfig(ctx, writeBucket, config, dependencies, writeConfigOptions.uncomment)
 	}
+	// Write out raw default config without comments using externalConfig{V1, V1Beta1}
+	return writeExternalConfig(ctx, writeBucket, version, config, dependencies)
+}
+
+func writeExternalConfig(
+	ctx context.Context,
+	writeBucket storage.WriteBucket,
+	version string,
+	config *Config,
+	dependencies []string,
+) error {
+	var name string
+	if config.ModuleIdentity != nil {
+		name = config.ModuleIdentity.IdentityString()
+	}
+	breakingConfig := config.Breaking
+	lintConfig := config.Lint
+	if version == V1Beta1Version {
```

Comment:
We should still write this so that the `version` must be either `v1beta1` or `v1` (instead of defaulting to `v1` when version is not `v1beta1`). I know we already validate it via `validateVersion` earlier, but this implicit default behavior is a little less obvious.

For consistency, we can adopt the `switch` approach that you see in the other configuration functions, like [getConfigForDataInternal](https://github.com/bufbuild/buf/blob/e4a9b7b053022398f70922a9d7241b822154fe5d/private/bufpkg/bufconfig/get.go#L88).

Similarly, we could write separate helper functions (e.g. `externalConfig{V1,V1Beta1}ForConfig`) so that the implementation reads similar to the `case` found in `getConfigForDataInternal`:

```go
case V1Beta1Version:
  var externalConfigV1Beta1 ExternalConfigV1Beta1
  if err := unmarshalStrict(data, &externalConfigV1Beta1); err != nil {
    return nil, err
  }
  return newConfigV1Beta1(externalConfigV1Beta1)
```